### DISCUSSION TITLE
auto LOD generations

### DISCUSSION THREAD

**Discussion by Samsy** [donmccurdy/glTF-Transform#1048] [General]
Hello ! 

I've been trying to setup an auto-LOD generation with gltf-transform using the mesh-simplifer : 

```
	await this.documentMax.transform(

			weld({tolerance: 0.0001}),

			simplify(

				{ simplifier: MeshoptSimplifier, ratio: 0.0, error: 1, lockBorder: true }
			)
		)
```

I did play with any weld and ratio / error configs for those, I wonder if there are some configs / anything that could even decimate / wrangle the geometry even more for real distant LOD's.

Like turning a 10k vertices geometry into 250 vertices kind of, a brutal simplification

**Comment by donmccurdy** [comment]
Hi @Samsy! Can you share a model for which the settings above are not decimating enough? 

To my understanding, the settings you're passing to `simplify(...)` would allow meshoptimizer to simplify the model as far as its _topology_ allows, but it can't simplify triangles that don't share vertices.

The pre-processing with `weld(...)` is intended to make sure vertices are shared as much as possible. A lower tolerance (say 0.01) might give you more aggressive welding, and you could also raise the `toleranceNormal` option which is new in v3.5:

```js
await document.transform(
  weld({tolerance: 0.1, toleranceNormal: Math.PI / 2}),
  simplify({ simplifier: MeshoptSimplifier, ratio: 0.001, error: 0.250 })
);
```

I'd expect that to give _very_ aggressive simplification for most models, and may need to see the model you're working with if that's not the case.

**Comment by Samsy** [comment]
Sorry for the late answer ! 

This is a sample of the model here : 

https://github.com/Samsy/glbs/blob/main/greenandheliport.glb
<img width="857" alt="Screenshot 2023-08-09 at 17 32 12" src="https://github.com/donmccurdy/glTF-Transform/assets/3786958/5649de7d-3582-42bc-909d-c22c5b86d0a9">


That transforms into : 

https://github.com/Samsy/glbs/blob/main/greenandheliportlods.glb
<img width="813" alt="Screenshot 2023-08-09 at 17 31 37" src="https://github.com/donmccurdy/glTF-Transform/assets/3786958/c0d9cfdb-bc96-4023-b291-668152c35011">

I wonder if there are some ways to get the result better, but I also guess this is working with the params you shared ! Some of the geometry is really wrangled, but that is the game of it :)

**Reply by donmccurdy** [reply]
If you're not in a rush, I'm hopeful that the next release of meshoptimizer (which should include attribute-aware simplification) might do better with this sort of geometry. Will test the model a bit and see if there are attributes that perform better now as well. With some very thin volumes in the scene above, the choice of weld tolerances and simplification error will need to be less aggressive than 10–25% of the bbox in any case.

If this is for a very distant LOD, then something like coarse voxel remeshing might also be a good technique to preserve rough appearance, but is not something I've yet implemented in this library.
